src/module.c

```diff
@@ -1003,16 +1012,39 @@ JL_DLLEXPORT jl_value_t *jl_module_names(jl_module_t *m, int all, int imported)
         jl_sym_t *asname = b->globalref->name;
         int hidden = jl_symbol_name(asname)[0]=='#';
         int main_public = (m == jl_main_module && !(asname == jl_eval_sym || asname == jl_include_sym));
-        if ((b->publicp ||
+        if ((jl_atomic_load_relaxed(&b->value) != (jl_value_t*)m) &&
+            ((exported ? b->exportp : b->publicp) ||
              (imported && b->imported) ||
+             (usings && _binding_is_from_explicit_using(b)) ||
              (jl_atomic_load_relaxed(&b->owner) == b && !b->imported && (all || main_public))) &&
             (all || (!b->deprecated && !hidden))) {
-            jl_array_grow_end(a, 1);
-            // n.b. change to jl_arrayset if array storage allocation for Array{Symbols,1} changes:
-            jl_array_ptr_set(a, jl_array_dim0(a)-1, (jl_value_t*)asname);
+            _append_symbol_to_bindings_array(a, asname);
         }
         table = jl_atomic_load_relaxed(&m->bindings);
     }
+}
+
+JL_DLLEXPORT jl_value_t *jl_module_names(jl_module_t *m, int all, int imported, int usings)
+{
+    jl_array_t *a = jl_alloc_array_1d(jl_array_symbol_type, 0);
+    JL_GC_PUSH1(&a);
+    // First, encode the module's name
+    _append_symbol_to_bindings_array(a, m->name);
+    _jl_module_names_into_array(a, m, all, imported, usings, /*exported*/0);
+    if (usings) {
+        // If `usings` is specified, traverse the list of `using`-ed modules and incorporate
+        // the names exported by those modules into the list.
+        for(int i=(int)m->usings.len-1; i >= 0; --i) {
+            jl_module_t *usinged = module_usings_getidx(m, i);
+            _jl_module_names_into_array(a, usinged, 0, 0, 0, /*exported*/1);
+            // Add the name of `usinged` itself, unless the user requested `all=true` and it's
+            // a submodule of `m`, since then its name would have already been added by
+            // `all=true`, since it's a binding in `m`.
+            if (!all || usinged->parent != m) {
```

Comment:
I'm not sure this is enough to avoid duplicates. For example the name might also be exported. We could also de-duplicate in the caller though.

Reply:
Yeah... I've been thinking about how to eliminate the duplication, and it seems it might be better to start by organizing the interface of `names`. With the current design and implementation, `all`/`imported` and `usings` are not necessarily orthogonal, so the duplication is inherent in the definitions.
By defining `names` as `names(x::Module; non_public::Bool=false, generated::Bool=false, imported::Bool=false, usings::Bool=false)` and refactoring the code, I think we can make the handling of such duplications more clear.

Reply:
The refactoring mentioned above involves more changes, so..., for now in this PR, I want to either 1) have `names` return a `Set` to eliminate duplicates on the `names` side (technically this is a breaking change?), or 2) add a warning about the possibility of duplicates in the `names` docstring. Which do you think is better?

Reply:
I've decided to go with the second option. Regarding the initial comment, I've found that this code itself is not necessary, so I will just remove it. The duplication issue remains (such as when there is a name conflicting with `using`-ed one), but at least there is a warning, so it shouldn't be a big problem.